Use OpenJDK11

belugabehr opened the pull request.

kgyrtkirk commented on Dockerfile:
```diff
@@ -20,6 +17,9 @@ COPY bin  /bin
 COPY tools/install_conf /tools/
 RUN /tools/install_conf
 
+RUN apt-get install -q -y binutils openjdk-11-jdk-headless
```
..please don't do this! follow the existing way of doing stuff... and move it back where it was - install_conf is a more  lightweight finalizer step ; moving this behind it will cause that every time in that file changes "java" is reinstalled...

we can't just switch to jdk11 only - there should be a way to still support java8 (and optionally java11) at the same time... in tools/install_sdk.bash  (which you have just removed) there were an approach to switch between version...(but it turned out that sdkman sucks in a container environment like this - so I replaced it with a plain jdk8)

I'm not sure about moving JAVA_HOME ...it have been appended to the initialization file "/etc/profile.d/confs.sh" - we should either use that or not at all; right now it only has env variables...but moving just one is a bad idea

belugabehr commented:
Thanks for the feedback.

I was trying to simplify things here.  I also quickly figured out that
sdkman is difficult to use in a container, hence I removed it altogether in
my PR.

In doing some research about installing Java, the consensus seemed to be
this was the way to go.  If we want different versions, probably should use
linux 'alternatives' command.

I'm not sure I understand the comment about having to reinstall java.  It's
not that lengthy and it makes the entire process much more straightforward
this way.

IMHO, with your background on this (since you authored it), you will
probably be spending more time directing me towards the solution you would
like to see than if you were just able to add JDK 11 support yourself.  Are
you able to help me out with this? Thanks!


On Tue, Jul 14, 2020, 8:16 AM Zoltan Haindrich <notifications@github.com>
wrote:

> *@kgyrtkirk* commented on this pull request.
> ------------------------------
>
> In Dockerfile
> <https://github.com/kgyrtkirk/hive-dev-box/pull/6#discussion_r454312032>:
>
> > @@ -20,6 +17,9 @@ COPY bin  /bin
>  COPY tools/install_conf /tools/
>  RUN /tools/install_conf
>
> +RUN apt-get install -q -y binutils openjdk-11-jdk-headless
>
> ..please don't do this! follow the existing way of doing stuff... and move
> it back where it was - install_conf is a more lightweight finalizer step ;
> moving this behind it will cause that every time in that file changes
> "java" is reinstalled...
>
> we can't just switch to jdk11 only - there should be a way to still
> support java8 (and optionally java11) at the same time... in
> tools/install_sdk.bash (which you have just removed) there were an approach
> to switch between version...(but it turned out that sdkman sucks in a
> container environment like this - so I replaced it with a plain jdk8)
>
> I'm not sure about moving JAVA_HOME ...it have been appended to the
> initialization file "/etc/profile.d/confs.sh" - we should either use that
> or not at all; right now it only has env variables...but moving just one is
> a bad idea
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kgyrtkirk/hive-dev-box/pull/6#pullrequestreview-448049749>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC766E6BWVNOCCDW5ETZ5QDR3RD77ANCNFSM4OYTBUCQ>
> .
>

belugabehr commented:
Also, we can have two different images with two different versions of Java and (if needed) two different build pipelines.  JDK 11 should be backwards compatible with JDK8 builds.  Not sure why we need to have some sort of dynamic feature for this since this will be part of the build pipeline.

belugabehr commented:
Thinking about it a little more, the root container should probably install maven, hadoop, etc. and then one child container for each version of Java (executor-jdk8, executor-jdk11, etc.) where it does the `yum install` and `ENV JAVA_HOME` statements.

kgyrtkirk commented:
bq. Thinking about it a little more, the root container should probably install maven, hadoop, etc. and then one child container for each version of Java (executor-jdk8, executor-jdk11, etc.) where it does the yum install and ENV JAVA_HOME statements.

maybe...thats one way to do it...but if you need 2 versions from something else you will end up creating a lot of images..

I don't want to break the existing containers - I'm leaving jdk8 as default; and added `sw java 11` to switch to jdk11

belugabehr commented:
That's great.  thanks for the help.  How would I trigger a build to use JDK 11?

kgyrtkirk commented:
over to the apache/hive#1241

belugabehr commented:
@kgyrtkirk Did you push `sw java 11` into this?  I don't see it in Github as of yet.

kgyrtkirk commented:
ugh...sorry forgot to push the changes... I just built the image and uploaded it
